pkg/controller/predicate/rolloutgate.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package predicate
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	k8spredicate "sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// RolloutGate allows reconcilers to select which resources they are enabled for, based
+// on features of the resource to-be-reconciled. This allows for a gated rollout of the
+// controller.
+type RolloutGate interface {
+	// ShouldReconcile returns true if the reconciler should be used to for the resource.
+	ShouldReconcile(o *unstructured.Unstructured) bool
+}
+
+// RolloutPredicate generates a controller-runtime predicate based on a RolloutGate.
+type RolloutPredicate struct {
+	gvk schema.GroupVersionKind
+	rg  RolloutGate
+	c   client.Client
+}
+
+var _ k8spredicate.Predicate = &RolloutPredicate{}
+
+func NewRolloutPredicate(c client.Client, gvk schema.GroupVersionKind, rg RolloutGate) *RolloutPredicate {
+	return &RolloutPredicate{
+		c:   c,
+		gvk: gvk,
+		rg:  rg,
+	}
+}
+
+func (p *RolloutPredicate) Create(e event.CreateEvent) bool {
+	obj, err := getUnstructuredObjWithGVK(p.c, e.Object, p.gvk)
+	if err != nil {
+		return false
+	}
+	return p.rg.ShouldReconcile(obj)
+}
+
+func (p *RolloutPredicate) Delete(e event.DeleteEvent) bool {
+	obj, err := getUnstructuredObjWithGVK(p.c, e.Object, p.gvk)
+	if err != nil {
+		return false
+	}
+	return p.rg.ShouldReconcile(obj)
+}
+
+func (p *RolloutPredicate) Update(e event.UpdateEvent) bool {
+	obj, err := getUnstructuredObjWithGVK(p.c, e.ObjectNew, p.gvk)
+	if err != nil {
+		return false
+	}
+	return p.rg.ShouldReconcile(obj)
+}
+
+func (p *RolloutPredicate) Generic(e event.GenericEvent) bool {
+	obj, err := getUnstructuredObjWithGVK(p.c, e.Object, p.gvk)
+	if err != nil {
+		return false
+	}
+	return p.rg.ShouldReconcile(obj)
```

Comment:
Nit: Feels like you could pass the event to ShouldReconcile?

Reply:
Ah I see, actually this makes sense because I think we're assuming that at some stage we will watch more intelligently and not the extra GET.  We could eliminate some of the duplication by having a helper method in `RolloutPredicate`, but ... not a blocker.

Reply:
Yeah, it felt like this was the least hassle way to do it, because the `UpdateEvent` passes in `ObjectNew` and `ObjectOld` (unlike all of the other events, which only pass in `Object`). So, there wasn't really any way to make a generic handler for every single event.